src/node_api.cc

```diff
@@ -3287,3 +3349,55 @@ napi_status napi_cancel_async_work(napi_env env, napi_async_work work) {
 
   return napi_clear_last_error(env);
 }
+
+NAPI_EXTERN napi_status napi_create_promise(napi_env env,
+                                            napi_value* deferred,
+                                            napi_value* promise) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(env, deferred);
+  CHECK_ARG(env, promise);
+
+  auto maybe = v8::Promise::Resolver::New(env->isolate->GetCurrentContext());
+  CHECK_MAYBE_EMPTY(env, maybe, napi_generic_failure);
+
+  auto v8_resolver = maybe.ToLocalChecked();
+
+  // Stick the V8 resolver into an empty shell object.
+  v8::Local<v8::Object> wrapper;
+  napi_status status = v8impl::EnvWrapperObject(env,
+                                                &env->deferred_template,
+                                                v8_resolver,
+                                                v8impl::napi_deferred_name,
+                                                &wrapper);
+  if (status != napi_ok) {
+    return napi_set_last_error(env, status);
+  }
+
+  *deferred = v8impl::JsValueFromV8LocalValue(wrapper);
+  *promise = v8impl::JsValueFromV8LocalValue(v8_resolver->GetPromise());
+  return GET_RETURN_STATUS(env);
+}
+
+NAPI_EXTERN napi_status napi_fulfill_deferred(napi_env env,
```

Comment:
Mh, just to be clear: [fulfilling](https://tc39.github.io/ecma262/#sec-fulfillpromise) and [rejecting](https://tc39.github.io/ecma262/#sec-promise-resolve-functions) are *different actions* in this context, and the terms are not interchangable.

Fulfilling a promise means putting it into a fulfilled state and triggering the first arguments of `.then()`.

Resolving a promise means either fulfilling it immediately, or letting the promise assume the state of another thenable object, which, in particular, could be a rejected promise (in which case resolving actually means rejecting).

What this method does is resolving the promise, so it should be called that.